nav2_controller/src/controller_server.cpp

```diff
@@ -246,7 +247,16 @@ ControllerServer::on_deactivate(const rclcpp_lifecycle::State & /*state*/)
   for (it = controllers_.begin(); it != controllers_.end(); ++it) {
     it->second->deactivate();
   }
-  costmap_ros_->deactivate();
+
+  /*
+   * The costmap is also a lifecycle node, so it may have already fired on_deactivate
+   * via rcl preshutdown cb. Let's not make an assumption about the order these fire...
+   */
+  if (costmap_ros_->get_current_state().id() ==
```

Comment:
In what situation would the costmap state differ from the server state? If none, then you can use the `state` given in this function to make that decision, no?

Also, the lifecycle_node.cpp entries are checking if the state is active to run deactivation on shutdown. So if the lifecycle manager already cycled it down, then it wouldn't trigger this -- so why the duplicate check here at all?

Reply:
Well, for a few reasons...   The first is that the rcl preshutdown callback may have fired on the costmap already, and in that case we don't want to call deactivate() on it here, since we needlessly get a log line if we do:

```
[187.382828] (nav2_costmap_2d) StdoutLine: {'line': b'22: [costmap_tests-5] [WARN] [1674969001.522910754] [rcl_lifecycle]: No transition matching 4 found for current state inactive\n'}
```

In that case we may ask why we call deactivate() on the costmap here at all, and the reason is that deactivate() can be called on the controller_server at any time, not just for shutdown. The current logic (prior to my change) is to also deactivate the costmap. My change ensures we still call it in the case that we aren't shutting down.

w.r.t. the lifecycle_node logic: same thing. We may not be shutting down here at all. This may be a deactivate transition that the user triggered themselves for some other reason.

Reply:
> rcl preshutdown callback may have fired on the costmap already

Is that because we register the preshutdown callback in the constructor? I would have assumed given the focus on determinism in ROS 2 that the execution of the preshutdown callbacks would be too. It looks like the costmap would always be before the controller/planner server since it completes its constructor while the planner/controller server's constructors are still crunching along. 

If that is true which would cause this issue, if we moved the costmap construction and node thread to the controller/planner server's `on_configure`, that would make it so that deterministically, those server's pre-shutdown callbacks should be called first (and thus we know preshutdown hasn't messed with the costmap state separately from the server's callbacks). Arguably, that's where those should be anyway since memory allocation should be in configuration. 

I think from the rest of what you mention, my suggestion is just an inversion of ordering which should resolve the warning you mention + not require checking the state in the server and push all that down to the already implemented logic in the lifecycle node base class.

Reply:
>  I would have assumed given the focus on determinism in ROS 2 that the execution of the preshutdown callbacks would be too.

Somewhat shockingly the callbacks are stored in an unordered set, which I believe means they will be executed in some arbitrary hash order.  That means they may even run in a different order between different platforms!

https://github.com/ros2/rclcpp/blob/rolling/rclcpp/include/rclcpp/context.hpp#L382

https://github.com/ros2/rclcpp/blob/rolling/rclcpp/src/rclcpp/context.cpp#L315

The reason it's shocking is that the on_shutdown (i.e. post-shutdown) callbacks are implemented the same way, and their docs specifically say the callbacks will be executed in the order added:

https://docs.ros2.org/foxy/api/rclcpp/classrclcpp_1_1Context.html#a64bce7cbf8b92f3e0283077f26facf9b

The preshutdown callbacks don't have docs that I have found, so technically nothing is in violation there, but it's certainly a surprise.  

In-fact I think the current code would work correctly (with the 'if' removed) if the callbacks *were* called in the order added, since super constructors run before child class constructors.

Seems like the docs are wrong? I'll float an issue to ROS2. But either way, my change needs to be this way for now...

Reply:
Filed: https://github.com/ros2/rclcpp/issues/2096

Reply:
I wish somebody would hire me to work on ROS2 + Nav2 so I could go whack a mole all this stuff!

Reply:
That's ridiculous, but thank you. I look at that and don't really see why that isn't a vector or even just a `set` anyway for how its called. 

 If you want to work on ROS 2 core, let me know, I can connect you :wink: